[CIVP-19308] New civis_file_to_table

zacharydfriedlander opened the pull request.

Commit:
Convert civis_file_to_table to use the new imports/files/csv and preprocess endpoints

Commit:
Add additional parameters now available via new endpoint

Commit:
Add docstring for new parameters

Commit:
Add additional logic to skip column detection when existing table columns can already be used

Commit:
Remove use of f-strings

Commit:
Update CHANGELOG

jkulzick commented on civis/io/_tables.py:
```diff
@@ -824,18 +826,21 @@ def csv_to_civis(filename, database, table, api_key=None, client=None,
 
 
-def civis_file_to_table(file_id, database, table, client=None,
+def civis_file_to_table(file_id_or_ids, database, table, client=None,
```
Why not just `file_ids`? `file_id_or_ids` is longer and is still ambiguous, meaning i'm still going to scratch my head and look at the docstring.

zacharydfriedlander replied:
Renamed.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -824,18 +826,21 @@ def csv_to_civis(filename, database, table, api_key=None, client=None,
                         diststyle=None, distkey=None,
                         sortkey1=None, sortkey2=None,
+                        primary_keys=None, last_modified_keys=None,
+                        escaped=False, execution="delayed",
```
I think `execution` should default to `immediate` because it is the default behavior that users will expect and actually having the default be `delayed` could be considered a breaking change since the behavior of syncing column statistics would no longer be blocking.

zacharydfriedlander replied:
That seems reasonable. I had selected `delayed` as the default in order to match the API.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -824,18 +826,21 @@ def csv_to_civis(filename, database, table, api_key=None, client=None,
                         sortkey1=None, sortkey2=None,
+                        primary_keys=None, last_modified_keys=None,
+                        escaped=False, execution="delayed",
+                        compression=None,
```
I don't think we actually need this parameter. I think it just complicates the interface. The file cleaner endpoint will tell us what the compression is.

zacharydfriedlander replied:
Removed.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -861,8 +866,31 @@ def civis_file_to_table(file_id, database, table, client=None,
     sortkey2 : str, optional
         The second column in a compound sortkey for the table.
+    primary_keys: list[str], optional
+        A list of the primary key column(s) of the destination table. If the
```
I think the criteria is not that the table exists, but that the table exists and the primary keys have been defined on the table. And that said, I think that is minutiae that is likely to just confuse users. I would be okay with just saying `If the import mode is "upsert", this field is required.`.

zacharydfriedlander replied:
I initially had copied this wording from the API docs - I've updated it for `primary_keys` and `last_modified_keys`.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -861,8 +866,31 @@ def civis_file_to_table(file_id, database, table, client=None,
+        A list of the primary key column(s) of the destination table. If the
+        destination table does not exist, and the import mode is
+        "upsert", this field is required.
+    last_modified_keys: list[str], optional
```
Same comment as `primary_keys`

jkulzick commented on civis/io/_tables.py:
```diff
@@ -861,8 +866,31 @@ def civis_file_to_table(file_id, database, table, client=None,
+        A boolean value indicating whether or not the source file has quotes
+        escaped with a backslash. Defaults to false.
+    execution: string, optional, default "delayed"
+        One of "delayed" and "immediate". if "immediate", refresh column
```
A couple small typos: `One of "delayed" _or_ "immediate". _I_f "immediate", refresh ... `

zacharydfriedlander replied:
Fixed.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -861,8 +866,31 @@ def civis_file_to_table(file_id, database, table, client=None,
+        escaped with a backslash. Defaults to false.
+    execution: string, optional, default "delayed"
+        One of "delayed" and "immediate". if "immediate", refresh column
+        statistics as part of the run. If "delayed", flag the table for a
+        deferred statistics update; column statistics may not be available
```
Another possible wording: `If "delayed", delay the refresh of column statistics for up to 24 hours.`

zacharydfriedlander replied:
Rephrased.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -861,8 +866,31 @@ def civis_file_to_table(file_id, database, table, client=None,
+        One of "delayed" and "immediate". if "immediate", refresh column
+        statistics as part of the run. If "delayed", flag the table for a
+        deferred statistics update; column statistics may not be available
+        until the next day. In addition, if existing_table_rows is "upsert",
```
We should probably stay consistent with how we refer to `"upsert"`. So either stick with referencing `existing_table_rows` or use the term `import mode`. I'm leaning towards the former.

zacharydfriedlander replied:
Agreed. Rephrased.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -861,8 +866,31 @@ def civis_file_to_table(file_id, database, table, client=None,
+        delayed executions move data from staging table to final table after a
+        brief delay, in order to accommodate multiple concurrent imports to the
+        same destination table.
+    compression: strip, optional
```
Per the above comment, I would remove this

zacharydfriedlander replied:
Removed

jkulzick commented on civis/io/_tables.py:
```diff
@@ -892,39 +920,106 @@ def civis_file_to_table(file_id, database, table, client=None,
+                raise CivisImportError('Mismatch between detected schemas:\n'
+                                       'Expected{}\nbut file {} has{}'.format(
+                                           table_columns,
+                                           output_file.name,
```
IMHO `output_file.name` will not be that helpful to the user or support. We should at least have the file id.

zacharydfriedlander replied:
Changed to use ids

jkulzick commented on civis/io/_tables.py:
```diff
@@ -892,39 +920,106 @@ def civis_file_to_table(file_id, database, table, client=None,
+        detected_info = client.files.get(output_file.object_id).detected_info
+        if not table_exists or existing_table_rows == 'drop':
+            file_columns = detected_info['tableColumns']
+            table_columns = table_columns or file_columns
```
I think this might be easier to understand if we separate the validation of the first `cleaning_futures` result from the others. It took me awhile to understand that the setting and processing of the various attributes differed (e.g., `headers` vs `table_columns`).

zacharydfriedlander replied:
The benefit of taking all the futures at once, and raising even on the first future, is that we get to catch inconsistencies between what the user has asked of us via provided parameters and the file the user has given us. I guess my question then is whether or not we trust our users.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -892,39 +920,106 @@ def civis_file_to_table(file_id, database, table, client=None,
+        if not table_exists or existing_table_rows == 'drop':
+            file_columns = detected_info['tableColumns']
+            table_columns = table_columns or file_columns
+            if table_columns != file_columns:
```
It seems like this should be converted to a utility function that compares two variables and raises the error if they don't match. Then we could call it for all of the below attributes we care about.

zacharydfriedlander replied:
That seems like needless indirection IMO - I think I'd rather do all of this inline.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -892,39 +920,106 @@ def civis_file_to_table(file_id, database, table, client=None,
+    except ValueError:
+        table_exists = False
+
+    # Use Preprocess endpoint to get the table columns
```
Should we move preprocessing to a private function? It would help with readability and unit testing.

zacharydfriedlander replied:
Done

jkulzick commented on civis/io/_tables.py:
```diff
@@ -892,39 +920,106 @@ def civis_file_to_table(file_id, database, table, client=None,
+            file_columns = detected_info['tableColumns']
+            table_columns = table_columns or file_columns
+            if table_columns != file_columns:
+                raise CivisImportError('Mismatch between detected schemas:\n'
```
Should we add something to the error message like `'All files must have the same format. Mismatch ...'`?

zacharydfriedlander replied:
Reworded

jkulzick commented on civis/io/_tables.py:
```diff
@@ -892,39 +920,106 @@ def civis_file_to_table(file_id, database, table, client=None,
+                                   'please ensure all imported files either '
+                                   'have a header or do not.')
+
+        if delimiter != detected_info['columnDelimiter']:
```
`delimiter` can also be passed in like `headers` so we should be consistent

zacharydfriedlander replied:
This had been set up in this way because `delimiter` has some validation set on it above. I've changed that validation to only apply if `delimiter` is passed in, and made the checking consistent here.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -892,39 +920,106 @@ def civis_file_to_table(file_id, database, table, client=None,
+        cleaner_job = client.files.post_preprocess_csv(
+            file_id=fid,
+            in_place=False,
+            detect_table_columns=(not table_exists
```
It looks like we reference these conditions twice (i.e., `not table_exists or existing_table_rows == 'drop'`). Perhaps we should encapsulate these conditions into a well-named variable, such as `table_columns_required` or `specify_table_columns`.

zacharydfriedlander replied:
Refactored.

Commit:
simplify primary_keys and last_modified_keys docstrings, and remove compresion parameter

Commit:
Add _replace_nulls function to handle empty column names from file cleaning

Commit:
Refactor messy boolean into need_table_columns

Commit:
Remove references to "import mode" in favor of existing_table_rows

Commit:
Rephrase execution docstring

Commit:
Check delimiter like we do headers

Commit:
Clean up error messages and use ids instead of names

Commit:
Change default execution to `immediate`

Commit:
Reword table columns error

Commit:
File_id_or_ids -> file_ids

Commit:
Move cleaning futures and processing cleaning to their own private helpers

Commit:
Update API spec

zacharydfriedlander commented:
I'm pretty sure something is off about how I've updated the cassettes - i've reverted those changes for now, and I'm going to keep investigating that.

Commit:
Update mock client to use non-channels spec

Commit:
Update civis_file_to_table specs

Commit:
Move mock client creation to test-level setup

Commit:
Add doesn't exist test

Commit:
Add a test for_process_cleaning_results

Commit:
add run_cleaning test

Commit:
just use fake file ids as fake job ids too

zacharydfriedlander commented:
So it looks like these tests are failing due to not having a real API key (and therefore being unable to construct an API client). I can confirm this by overriding my local API key with a nonsense one - my question is, why aren't these tests getting caught by cassettes? Maybe due to the updated API spec?

Commit:
Remove unneeded todos

zacharydfriedlander commented:
So according to a `git bisect`, [` a7ce5bf16c84b68ac1ba40fa407c2df555067a44`](https://github.com/zacharydfriedlander/civis-python/commit/a7ce5bf16c84b68ac1ba40fa407c2df555067a44) is the bad commit.

I've done some more experimentation, and it looks like the issue occurs if _any_ `setUp` method is defined, even a blank one. Why this is the case, I have no idea.

Commit:
Convert setUp function into an autoused pytest fixture

jkulzick commented on civis/io/_tables.py:
```diff
@@ -1094,3 +1140,107 @@ def split_schema_tablename(table):
     return tuple(schema_name_tup)
+
+
+def _replace_nulls(column_list):
```
I think a more explicit name would be good. Perhaps something like `_replace_null_column_names`

jkulzick commented on civis/io/_tables.py:
```diff
@@ -1094,3 +1140,107 @@ def split_schema_tablename(table):
+            first_completed.run_id
+        )[0]
+    detected_info = client.files.get(output_file.object_id).detected_info
+    if need_table_columns:
```
Perhaps move the default assignment here from line 1185 as well so the context is not spread out

jkulzick commented on civis/io/_tables.py:
```diff
@@ -1094,3 +1140,107 @@ def split_schema_tablename(table):
+        headers = detected_info['includeHeader']
+    if delimiter is None:
+        delimiter = detected_info['columnDelimiter']
+    if compression is None:
```
Same comment as `table_columns`

zacharydfriedlander replied:
That first initialization was entirely unnecessary - removed.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -1094,3 +1140,107 @@ def split_schema_tablename(table):
+        )[0]
+        detected_info = client.files.get(output_file.object_id).detected_info
+        if need_table_columns:
+            file_columns = _replace_nulls(detected_info['tableColumns'])
```
Another approach would be to only call `_replace_nulls` at the end after validating all of the futures

zacharydfriedlander replied:
Done

jkulzick commented on civis/io/_tables.py:
```diff
@@ -1094,3 +1140,107 @@ def split_schema_tablename(table):
+        detected_info = client.files.get(output_file.object_id).detected_info
+        if need_table_columns:
+            file_columns = _replace_nulls(detected_info['tableColumns'])
+            if table_columns != file_columns:
```
I find the code for these comparisons to be a bit repetitive. The benefit is that you can highly customize the error message (although you could pass in an error message prefix to some common function) and the downside is that there is more code to maintain. I don't have any strong opinions either way.

zacharydfriedlander replied:
I don't mind the repetition, and I'd rather reduce the amount of indirection needed to understand what's happening or where an error comes from.

jkulzick commented on civis/io/_tables.py:
```diff
@@ -1094,3 +1140,107 @@ def split_schema_tablename(table):
+    detected_info = client.files.get(output_file.object_id).detected_info
+    if need_table_columns:
+        table_columns = _replace_nulls(detected_info['tableColumns'])
+    if headers is None:
```
Strictly speaking following the validation laid out we would need to validate both `headers` and `delimiter` in the first completed job against the values provided when they are provided.

zacharydfriedlander replied:
Added

jkulzick commented on civis/io/_tables.py:
```diff
@@ -1094,3 +1140,107 @@ def split_schema_tablename(table):
+    done, still_going = concurrent.futures.wait(
+        cleaning_futures, return_when=concurrent.futures.FIRST_COMPLETED
+    )
+    assert len(done) == 1
```
I'm curious why this is here? If we are not confident that `concurrent.futures.FIRST_COMPLETED` works should we use something else?

zacharydfriedlander replied:
I think I had this here as a guard more against Platform completing two jobs at the same time than against `concurrent.futures.wait`. However, I think the odds of that happening (especially in any sort of reproducible way) are slim. Removed.

jkulzick commented on civis/tests/test_io.py:
```diff
@@ -196,66 +199,244 @@ def test_csv_to_civis(self, m_civis_file_to_table, m_file_to_civis,
-        self.mock_client.jobs.post_runs.assert_called_once_with(mock_import_id)
+
+    @pytest.mark.process_cleaning_results
+    def test_process_cleaning_results(self, _m_get_api_spec):
```
We should probably have some tests that cover the scenarios when we raise a `CivisImportError`

zacharydfriedlander replied:
Added

jkulzick commented on civis/io/_tables.py:
```diff
@@ -1094,3 +1140,107 @@ def split_schema_tablename(table):
+                                           file_columns)
+                                       )
+        if headers != detected_info['includeHeader']:
+            raise CivisImportError('Mismatch between detected headers - '
```
There are slightly different amounts of information in all 4 of these error messages. In `table_columns` and `delimiter` the expected and actual results are provided along with the offending file id. In `headers` and `compression` these details are not provided.

zacharydfriedlander replied:
Added for `compression`. However, for headers the exact phrasing IMO is harder because it's a boolean. Open to suggestions for how best to word that error.

Commit:
rename _replace_nulls to _replace_null_column_names and move call to the end

Commit:
refactor getting the first results out of the future

- Move tuple unpacking in-line
- remove assertion on len(done)
-

Commit:
Refactor initial future checsk

Commit:
Enhance later error messages with more info

Commit:
Remove compression initialization in favor of just pulling from detected info

Commit:
Add test for errors, and fix a bug therein

jkulzick reviewed (approved):
LGTM

zacharydfriedlander commented:
So it looks like there's an issue with multi-file uploads where `concurrent.futures.wait` will return more than one future in the `done` tuple on that `FIRST_COMPLETED` setting. I am going to refactor that code to be able to account for it.

Commit:
Replace assumption of a single done tuple in favor of picking one to be done first via set.pop(), and chaining the rest with the still-running futures

Commit:
Remove mocking of concurrent.futures.wait in test_io

Commit:
Remove use of itertools.chain in favor of just or-ing sets together.

Commit:
Additional bugfix around _replace_null_column_names unguarded by need_table_columns

jkulzick reviewed (approved):
LGTM
